espnet2/speechlm/core_lm/ar_delay.py

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/env python3
+
+# Copyright 2024 Jinchuan Tian
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+# Implementation of Delay architecture: https://arxiv.org/pdf/2306.05284
+
+import logging
+from typing import Dict, Tuple
+
+import torch
+
+from espnet2.speechlm.core_lm.abs_core_lm import SpeechLMInferenceOptions
+from espnet2.speechlm.core_lm.ar_parallel import ARParallelLM
+from espnet2.speechlm.inference_utils import AbsInferenceConfig
+from espnet2.speechlm.net_utils import (
+    install_continuous_features,
+    logits_to_tokens,
+    modality_index_to_mask,
+)
+
+
+class ARDelayLM(ARParallelLM):
+    def forward(
+        self,
+        dec_seq: torch.Tensor,
+        loss_mask: torch.Tensor = None,
+        conti_feats: list = None,
+    ) -> Tuple[torch.Tensor, Dict, torch.Tensor]:
+        """ARDelayLM forward for training.
+
+        This forward function is very similar to that of ARParallelLM only except the
+        delay interleave pattern
+
+        Args:
+            dec_seq (LongTensor): Batch of decoder sequences (B, T, nq).
+            loss_mask (LongTensor): Lengths of condition part in dec_seq (B, T, nq).
+        """
+        dec_seq_delay, loss_mask = self.delay_interleave(
+            dec_seq=dec_seq, loss_mask=loss_mask
+        )
+
+        return super().forward(
+            dec_seq=dec_seq_delay,
+            loss_mask=loss_mask,
+            conti_feats=conti_feats,
+        )
+
+    def delay_interleave(
+        self,
+        dec_seq: torch.Tensor,
+        loss_mask: torch.Tensor = None,
+        pad: int = 0,
+    ):
+        B, T, nq = dec_seq.size()
+        retval = (
+            torch.ones(
+                (B, T + self.nq - 1, nq), dtype=dec_seq.dtype, device=dec_seq.device
+            )
+            * pad
+        )
+        ret_loss_mask = retval.clone().detach() if loss_mask is not None else None
+
+        for n in range(self.nq):
+            retval[:, n : n + T, n] = dec_seq[:, :, n]
+            if ret_loss_mask is not None:
+                ret_loss_mask[:, n : n + T, n] = loss_mask[:, :, n]
+
+        retval = retval[:, : -(self.nq - 1)]
+        if ret_loss_mask is not None:
+            ret_loss_mask = ret_loss_mask[:, : -(self.nq - 1)]
+
+        return retval, ret_loss_mask
+
+    def inverse_delay_interleave(
```

Comment:
since delay-interleave and its reverse are deterministic, I think we should have a quick unit test to verify the input/output are properly shifted.